doc/gvpe.5
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` 
.    ds C' 
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GVPE 5"
.TH GVPE 5 "2008-09-01" "2.2" "GNU Virtual Private Ethernet"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
GNU\-VPE \- Overview of the GNU Virtual Private Ethernet suite.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1GVPE\s0 is a suite designed to provide a virtual private network for multiple
nodes over an untrusted network. This document first gives an introduction
to VPNs in general and then describes the specific implementation of \s-1GVPE\s0.
.Sh "\s-1WHAT\s0 \s-1IS\s0 A \s-1VPN\s0?"
.IX Subsection "WHAT IS A VPN?"
\&\s-1VPN\s0 is an acronym, it stands for:
.IP "Virtual" 4
.IX Item "Virtual"
Virtual means that no physical network is created (of course), but a
network is \fIemulated\fR by creating multiple tunnels between the member
nodes by encapsulating and sending data over another transport network.
.Sp
Usually the emulated network is a normal \s-1IP\s0 or Ethernet, and the transport
network is the Internet. However, using a \s-1VPN\s0 system like \s-1GVPE\s0 to connect
nodes over other untrusted networks such as Wireless \s-1LAN\s0 is not uncommon.
.IP "Private" 4
.IX Item "Private"
Private means that non-participating nodes cannot decode (\*(L"sniff)\*(R" nor
inject (\*(L"spoof\*(R") packets. This means that nodes can be connected over
untrusted networks such as the public Internet without fear of being
eavesdropped while at the same time being able to trust data sent by other
nodes.
.Sp
In the case of \s-1GVPE\s0, even participating nodes cannot sniff packets
send to other nodes or spoof packets as if sent from other nodes, so
communications between any two nodes is private to those two nodes.
.IP "Network" 4
.IX Item "Network"
Network means that more than two parties can participate in the network,
so for instance it's possible to connect multiple branches of a company
into a single network. Many so-called \*(L"\s-1VPN\s0\*(R" solutions only create
point-to-point tunnels, which in turn can be used to build larger
networks.
.Sp
\&\s-1GVPE\s0 provides a true multi-point network in which any number of nodes (at
least a few dozen in practise, the theoretical limit is 4095 nodes) can
participate.
.Sh "\s-1GVPE\s0 \s-1DESIGN\s0 \s-1GOALS\s0"
.IX Subsection "GVPE DESIGN GOALS"
.IP "\s-1SIMPLE\s0 \s-1DESIGN\s0" 4
.IX Item "SIMPLE DESIGN"
Cipher, \s-1HMAC\s0 algorithms and other key parameters must be selected
at compile time \- this makes it possible to only link in algorithms
you actually need. It also makes the crypto part of the source very
transparent and easy to inspect, and last not least this makes it possible
to hardcode the layout of all packets into the binary. \s-1GVPE\s0 goes a step
further and internally reserves blocks of the same length for all packets,
which virtually removes all possibilities of buffer overflows, as there is
only a single type of buffer and it's always of fixed length.
.IP "\s-1EASY\s0 \s-1TO\s0 \s-1SETUP\s0" 4
.IX Item "EASY TO SETUP"
A few lines of config (the config file is shared unmodified between all
hosts) and a single run of \f(CW\*(C`gvpectrl\*(C'\fR to generate the keys suffices to
make it work.
.IP "MAC-BASED \s-1SECURITY\s0" 4
.IX Item "MAC-BASED SECURITY"
Since every host has it's own private key, other hosts cannot spoof
traffic from this host. That makes it possible to filter packet by \s-1MAC\s0
address, e.g. to ensure that packets from a specific \s-1IP\s0 address come, in
fact, from a specific host that is associated with that \s-1IP\s0 and not from
another host.
.SH "PROGRAMS"
.IX Header "PROGRAMS"
Gvpe comes with two programs: one daemon (\f(CW\*(C`gvpe\*(C'\fR) and one control program
(\f(CW\*(C`gvpectrl\*(C'\fR).
.IP "gvpectrl" 4
.IX Item "gvpectrl"
This program is used to generate the keys, check and give an overview of of the
configuration and to control the daemon (restarting etc.).
.IP "gvpe" 4
.IX Item "gvpe"
This is the daemon used to establish and maintain connections to the other
network nodes. It should be run on the gateway of each \s-1VPN\s0 subnet.
.SH "COMPILETIME CONFIGURATION"
.IX Header "COMPILETIME CONFIGURATION"
Please have a look at the \f(CW\*(C`gvpe.osdep(5)\*(C'\fR manpage for platform-specific
information.
.PP
Gvpe hardcodes most encryption parameters. While this reduces flexibility,
it makes the program much simpler and helps making buffer overflows
impossible under most circumstances.
.PP
Here are a few recipes for compiling your gvpe, showing the extremes
(fast, small, insecure \s-1OR\s0 slow, large, more secure), between which you
should choose:
.Sh "\s-1AS\s0 \s-1LOW\s0 \s-1PACKET\s0 \s-1OVERHEAD\s0 \s-1AS\s0 \s-1POSSIBLE\s0"
.IX Subsection "AS LOW PACKET OVERHEAD AS POSSIBLE"
.Vb 1
\&   ./configure \-\-enable\-hmac\-length=4 \-\-enable\-rand\-length=0
.Ve
.PP
Minimize the header overhead of \s-1VPN\s0 packets (the above will result in
only 4 bytes of overhead over the raw ethernet frame). This is a insecure
configuration because a \s-1HMAC\s0 length of 4 makes collision attacks based on
the birthday paradox pretty easy.
.Sh "\s-1MINIMIZE\s0 \s-1CPU\s0 \s-1TIME\s0 \s-1REQUIRED\s0"
.IX Subsection "MINIMIZE CPU TIME REQUIRED"
.Vb 1
\&   ./configure \-\-enable\-cipher=bf \-\-enable\-digest=md4
.Ve
.PP
Use the fastest cipher and digest algorithms currently available in
gvpe. \s-1MD4\s0 has been broken and is quite insecure, though, so using another
digest algorithm is recommended.
.Sh "\s-1MAXIMIZE\s0 \s-1SECURITY\s0"
.IX Subsection "MAXIMIZE SECURITY"
.Vb 1
\&   ./configure \-\-enable\-hmac\-length=16 \-\-enable\-rand\-length=8 \-\-enable\-digest=sha1
.Ve
.PP
This uses a 16 byte \s-1HMAC\s0 checksum to authenticate packets (I guess 8\-12
would also be pretty secure ;) and will additionally prefix each packet
with 8 bytes of random data. In the long run, people should move to
\&\s-1SHA\-256\s0 and beyond).
.PP
In general, remember that \s-1AES\-128\s0 seems to be as secure but faster than
\&\s-1AES\-192\s0 or \s-1AES\-256\s0, more randomness helps against sniffing and a longer
\&\s-1HMAC\s0 helps against spoofing. \s-1MD4\s0 is a fast digest, \s-1SHA1\s0, \s-1RIPEMD160\s0, \s-1SHA256\s0
are consecutively better, and Blowfish is a fast cipher (and also quite
secure).
.SH "HOW TO SET UP A SIMPLE VPN"
.IX Header "HOW TO SET UP A SIMPLE VPN"
In this section I will describe how to get a simple \s-1VPN\s0 consisting of
three hosts up and running.
.Sh "\s-1STEP\s0 1: configuration"
.IX Subsection "STEP 1: configuration"
First you have to create a daemon configuration file and put it into the
configuration directory. This is usually \f(CW\*(C`/etc/gvpe\*(C'\fR, depending on how you
configured gvpe, and can be overwritten using the \f(CW\*(C`\-c\*(C'\fR command line switch.
.PP
Put the following lines into \f(CW\*(C`/etc/gvpe/gvpe.conf\*(C'\fR:
.PP
.Vb 3
\&   udp\-port = 50000 # the external port to listen on (configure your firewall)
\&   mtu = 1400       # minimum MTU of all outgoing interfaces on all hosts
\&   ifname = vpn0    # the local network device name
\&
\&   node = first     # just a nickname
\&   hostname = first.example.net # the DNS name or IP address of the host
\&
\&   node = second
\&   hostname = 133.55.82.9
\&
\&   node = third
\&   hostname = third.example.net
.Ve
.PP
The only other file necessary is the \f(CW\*(C`if\-up\*(C'\fR script that initializes the
virtual ethernet interface on the local host. Put the following lines into
\&\f(CW\*(C`/etc/gvpe/if\-up\*(C'\fR and make it executable (\f(CW\*(C`chmod 755 /etc/gvpe/if\-up\*(C'\fR):
.PP
.Vb 6
\&   #!/bin/sh
\&   ip link set $IFNAME address $MAC mtu $MTU up
\&   [ $NODENAME = first  ] && ip addr add 10.0.1.1 dev $IFNAME
\&   [ $NODENAME = second ] && ip addr add 10.0.2.1 dev $IFNAME
\&   [ $NODENAME = third  ] && ip addr add 10.0.3.1 dev $IFNAME
\&   ip route add 10.0.0.0/16 dev $IFNAME
.Ve
.PP
This script will give each node a different \s-1IP\s0 address in the \f(CW\*(C`10.0/16\*(C'\fR
network.  The internal network (if gvpe runs on a router) should then be
set to a subset of that network, e.g.  \f(CW\*(C`10.0.1.0/24\*(C'\fR on node \f(CW\*(C`first\*(C'\fR,
\&\f(CW\*(C`10.0.2.0/24\*(C'\fR on node \f(CW\*(C`second\*(C'\fR, and so on.
.PP
By enabling routing on the gateway host that runs \f(CW\*(C`gvpe\*(C'\fR all nodes will
be able to reach the other nodes. You can, of course, also use proxy \s-1ARP\s0
or other means of pseudo-bridging, or (best) full routing \- the choice is
yours.
.Sh "\s-1STEP\s0 2: create the \s-1RSA\s0 key pairs for all hosts"
.IX Subsection "STEP 2: create the RSA key pairs for all hosts"
Run the following command to generate all key pairs for all nodes (that
might take a while):
.PP
.Vb 1
\&   gvpectrl \-c /etc/gvpe \-g
.Ve
.PP
This command will put the public keys into \f(CW\*(C`/etc/gvpe/pubkeys/\f(CInodename\f(CW\*(C'\fR and the private keys into \f(CW\*(C`/etc/gvpe/hostkeys/\f(CInodename\f(CW\*(C'\fR.
.Sh "\s-1STEP\s0 3: distribute the config files to all nodes"
.IX Subsection "STEP 3: distribute the config files to all nodes"
Now distribute the config files and private keys to the other nodes. This
should be done in two steps, since only the private keys meant for a node
should be distributed (so each node has only it's own private key).
.PP
The example uses rsync-over-ssh
.PP
First all the config files without the hostkeys should be distributed:
.PP
.Vb 3
\&   rsync \-avzessh /etc/gvpe first.example.net:/etc/. \-\-exclude hostkeys
\&   rsync \-avzessh /etc/gvpe 133.55.82.9:/etc/. \-\-exclude hostkeys
\&   rsync \-avzessh /etc/gvpe third.example.net:/etc/. \-\-exclude hostkeys
.Ve
.PP
Then the hostkeys should be copied:
.PP
.Vb 3
\&   rsync \-avzessh /etc/gvpe/hostkeys/first  first.example.net:/etc/hostkey
\&   rsync \-avzessh /etc/gvpe/hostkeys/second 133.55.82.9:/etc/hostkey
\&   rsync \-avzessh /etc/gvpe/hostkeys/third  third.example.net:/etc/hostkey
.Ve
.PP
You should now check the configuration by issuing the command \f(CW\*(C`gvpectrl \-c
/etc/gvpe \-s\*(C'\fR on each node and verify it's output.
.Sh "\s-1STEP\s0 4: starting gvpe"
.IX Subsection "STEP 4: starting gvpe"
You should then start gvpe on each node by issuing a command like:
.PP
.Vb 1
\&   gvpe \-D \-l info first # first is the nodename
.Ve
.PP
This will make the gvpe daemon stay in foreground. You should then see
\&\*(L"connection established\*(R" messages. If you don't see them check your
firewall and routing (use tcpdump ;).
.PP
If this works you should check your networking setup by pinging various
endpoints.
.PP
To make gvpe run more permanently you can either run it as a daemon (by
starting it without the \f(CW\*(C`\-D\*(C'\fR switch), or, much better, from your inittab
or equivalent. I use a line like this on all my systems:
.PP
.Vb 1
\&   t1:2345:respawn:/opt/gvpe/sbin/gvpe \-D \-L first >/dev/null 2>&1
.Ve
.Sh "\s-1STEP\s0 5: enjoy"
.IX Subsection "STEP 5: enjoy"
\&... and play around. Sending a \-HUP (\f(CW\*(C`gvpectrl \-kHUP\*(C'\fR) to the daemon
will make it try to connect to all other nodes again. If you run it from
inittab, as is recommended, \f(CW\*(C`gvpectrl \-k\*(C'\fR (or simply \f(CW\*(C`killall gvpe\*(C'\fR) will
kill the daemon, start it again, making it read it's configuration files
again.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIgvpe.osdep\fR\|(5) for OS-dependent information, \fIgvpe.conf\fR\|(5), \fIgvpectrl\fR\|(8),
and for a description of the transports, protocol, and routing algorithm,
\&\fIgvpe.protocol\fR\|(7).
.PP
The \s-1GVPE\s0 mailing list, at <http://lists.schmorp.de/>, or
\&\f(CW\*(C`gvpe@lists.schmorp.de\*(C'\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
Marc Lehmann <gvpe@schmorp.de>
.SH "COPYRIGHTS AND LICENSES"
.IX Header "COPYRIGHTS AND LICENSES"
\&\s-1GVPE\s0 itself is distributed under the \s-1GENERAL\s0 \s-1PUBLIC\s0 \s-1LICENSE\s0 (see the file
\&\s-1COPYING\s0 that should be part of your distribution).
.PP
In some configurations it uses modified versions of the tinc vpn suite,
which is also available under the \s-1GENERAL\s0 \s-1PUBLIC\s0 \s-1LICENSE\s0.


NEWS
GVPE NEWS

2.21 Wed Sep  3 06:56:27 CEST 2008
	- add missing ev++.h include header to tarball, which everybody
          who tested it apparently had in their include path :(. Caught
          by Karl Kleinpaste.

2.2  Mon Sep  1 06:28:09 CEST 2008
	- protocol version 0.1, compatible with older releases,
          but upgrade is recommended to due changed ondemand behaviour.
        - new per-node options max-ttl and max-queue.
        - convert from iom.C to libev, a high-performance event loop
          (http://software.schmorp.de/pkg/libev).
        - tcp connections were leaking in some cases.
        - retry more aggressively (once/s) to establish a connection if
          new packets arrive for it.
        - save a lot of setsockopt calls when the tos doesn't change.
        - honor disabled even on initial connect attempt.
        - changed callback mechanism to be slightly less portable
          but more efficient mechanism (standards-compliant c++ compilers
          should work).
        - increased receive window positive size, to allow for massive
          packet loss due to occasional longer drop-outs.
        - send RST when a positive window size violation is detected, but
          not in other cases, to reconnect more quickly.
        - upgraded liblzf to version 3.4.
        - dropped -fno-exceptions due to ev++.h using it.
        - node-up/down scripts are now run in sequence.
        - new -q switch for gvpectrl, for when you run it often.
        - work around the horribly inconsistent, ad-hoc, ever-changing
          and broken texinfo syntax. YMMV. avoid texinfo.
        - keepalive is more aggressive now, sensding ping's every 3 seconds
          and killing the conenction after 15 seconds.
        - bugfixes.

2.01 Thu Mar 29 19:26:04 CEST 2007
	- protocol version 0.1, compatible with older releases.
        - bugfix of callback.h, might have cause callback return values to
          be corrupted on architectures like sparc before.
        - dns transport retries more aggressively.
        - updated documentation, improved dns transport reliability
          and throughput.
        - added experimental support for sha256 and sha512 digests.

2.0 Mon Dec  5 13:59:26 CET 2005
	- protocol version 0.1, compatible with older releases.
        - implement allow-direct, deny-direct node config statements.
        - implemented != for sockinfo. This fixes a bug where
          gvpe sent packets to the old ip address of another host
          even though it had received packets from it's new address.
          This only causes problems if you forget to -HUP your gvpe after
          your ip address changed, which is *required*.
        - sets close-on-exec flag on tcp connections. This fixes a bug
          where child processes kept tcp connections open and caused
          connections to fail when only one side can connect.
        - fixed a bug in receive sequence checking that made gvpe
          accept out-of-window packets in most cases.
        - tighter limit for the maximum sequence # to avoid overflow
          conditions + allow more headroom for packet reordering.
        - replace some asserts that trapped config mismatches by
          more useful log messages.
        - fix spurious extra newline in some log messages.

1.9 Tue Apr 19 06:21:50 CEST 2005
	- protocol version 0.1, compatible with older releases.
        - WARNING: this version checks the return value of if-up etc.
          scripts and exits on failure.
        - IMPORTANT: run if-up/node-up etc. scripts via /bin/sh.
        - IMPORTANT: interface initialization (MAC, MTU) is now done
          automatically in most configurations.
        - options can now be specified on the gvpe command line, too.
        - make some DNS transport values configurable and document them.
        - improved OS specific information in gvpe.osdep(5).
        - new tap device type "native/darwin", that supports the tap
          driver available for darwin (thanks to matthew mead who tested
          it out with me in a long session). tincd/darwin is still available.
        - new device type "tincd/bsd", which is a newer version of the
          *bsd-drivers taken from tinc.
        - fixed a bug in relying on the order of global construction
          when tcp transport was enabled. The fix makes it use no
          cpu time unless it is in use, too.
        - information about other nodes is now available to if-up etc. scripts.
        - the value of the config variable if-up-data is passed to the if-up
          etc. scripts.
	- skip unparsable config lines with a warning instead of stopping
          parsing and continuing with a certainly unusable config.

1.8 Fri Mar 18 00:58:55 CET 2005
	- protocol version 0.1, compatible with older releases.
	- enable-udp = yes is now default only when no other protocols
          are enabled. otherwise it is disabled unless explicitly enabled.
        - implemented dns tunneling (experimental now and in the future).
        - remove support for pre-release version protocol.
        - updated tincd drivers (rev 1433), added uml_socket driver, documented
          tincd drivers a bit better.
        - document icmp configuration values.
        - document transport protocols in gvpe.protocol(7).
        - remove unused ChangeLog file.
        - created a mailinglist at gvpe@lists.schmorp.de.
        - added an exemption to allow distribution of binaries linked against
          OpenSSL, as suggested by Guus Sliepen (author of tinc). No
          GNUTLS conversion in sight.
        - some portability fixes with respect to --disable-nls.

1.7 Tue Feb 22 23:58:59 CET 2005
	- protocol version 0.1, compatible with older releases.
	- first gnu release.
        - documented the special value 1 for router-priority.
        - renamed vped => gvpe and vpectrl => gvpectrl, as well as
          vped.conf => gvpe.conf.
	- new per-node option "max-retry".

src/gvpe.C
/*
    gvpe.C -- the main file for gvpe
    Copyright (C) 1998-2002 Ivo Timmermans <ivo@o2w.nl>
                  2000-2002 Guus Sliepen <guus@sliepen.eu.org>
                  2003-2008 Marc Lehmann <gvpe@schmorp.de>
 
    This file is part of GVPE.

    GVPE is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation; either version 3 of the License, or (at your
    option) any later version.
   
    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
    Public License for more details.
   
    You should have received a copy of the GNU General Public License along
    with this program; if not, see <http://www.gnu.org/licenses/>.
   
    Additional permission under GNU GPL version 3 section 7
   
    If you modify this Program, or any covered work, by linking or
    combining it with the OpenSSL project's OpenSSL library (or a modified
    version of that library), containing parts covered by the terms of the
    OpenSSL or SSLeay licenses, the licensors of this Program grant you
    additional permission to convey the resulting work.  Corresponding
    Source for a non-source form of such a combination shall include the
    source code for the parts of OpenSSL used as well as that of the
    covered work.
*/

#include "config.h"

#include <cstdio>
#include <cstring>
#include <cstdlib>

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
#include <signal.h>
#include <termios.h>

#if HAVE_SYS_MMAN_H
# include <sys/mman.h>
#endif

#include <openssl/err.h>
#include <openssl/rand.h>

#include "gettext.h"
#include "pidfile.h"

#include "conf.h"
#include "slog.h"
#include "util.h"
#include "vpn.h"
#include "ev_cpp.h"

static loglevel llevel = L_NONE;

/* If nonzero, display usage information and exit. */
static int show_help;

/* If nonzero, print the version on standard output and exit.  */
static int show_version;

/* If nonzero, disable swapping for this process. */
static int do_mlock = 0;

/* If zero, don't detach from the terminal. */
static int do_detach = 1;

static struct option const long_options[] =
    {
      {"config", required_argument, NULL, 'c'},
      {"help", no_argument, &show_help, 1},
      {"version", no_argument, &show_version, 1},
      {"no-detach", no_argument, &do_detach, 0},
      {"log-level", required_argument, NULL, 'l'},
      {"mlock", no_argument, &do_mlock, 1},
      {NULL, 0, NULL, 0}
    };

static void
usage (int status)
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help\' for more information.\n"), get_identity ());
  else
    {
      printf (_("Usage: %s [option]... NODENAME\n\n"), get_identity ());
      printf (_
              ("  -c, --config=DIR           Read configuration options from DIR.\n"
               "  -D, --no-detach            Don't fork and detach.\n"
               "  -l, --log-level=LEVEL      Set logging level (info, notice, warn are common).\n"
               "  -L, --mlock                Lock tinc into main memory.\n"
               "      --help                 Display this help and exit.\n"
               "      --version              Output version information and exit.\n\n"));
      printf (_("Report bugs to <gvpe@schmorp.de>.\n"));
    }

  exit (status);
}

void
parse_options (int argc, char **argv, char **envp)
{
  int r;
  int option_index = 0;

  while ((r = getopt_long (argc, argv, "c:DLl:", long_options, &option_index)) != EOF)
    {
      switch (r)
        {
        case 0:		/* long option */
          break;

        case 'c':		/* config file */
          confbase = strdup (optarg);
          break;

        case 'D':		/* no detach */
          do_detach = 0;
          break;

        case 'L':		/* lock into memory */
          do_mlock = 1;
          break;

        case 'l':		/* inc debug level */
          {
            llevel = string_to_loglevel (optarg);

            if (llevel == L_NONE)
              slog (L_WARN, "'%s': %s", optarg, UNKNOWN_LOGLEVEL);
          }
          break;

        case '?':
          usage (1);

        default:
          break;
        }
    }
}

/*
  Close network connections, and terminate neatly
*/
void cleanup_and_exit(int c)
{
  network.shutdown_all ();

  if (conf.pidfilename)
    remove_pid (conf.pidfilename);

  slog (L_INFO, _("terminating with exit code %d"), c);

  exit (c);
}

/*
  Signal handlers.
*/
RETSIGTYPE
sigterm_handler (int a)
{
  network.events |= vpn::EVENT_SHUTDOWN;
  network.event.start ();
}

RETSIGTYPE
sighup_handler (int a)
{
  network.events |= vpn::EVENT_RECONNECT;
  network.event.start ();
}

RETSIGTYPE
sigusr1_handler (int a)
{
  network.dump_status ();
}

RETSIGTYPE
sigusr2_handler (int a)
{
}

void
setup_signals (void)
{
  struct sigaction act;

  sigfillset (&act.sa_mask);
  act.sa_flags = 0;

  act.sa_handler = sighup_handler;  sigaction (SIGHUP , &act, NULL);
  act.sa_handler = sigusr1_handler; sigaction (SIGUSR1, &act, NULL);
  act.sa_handler = sigusr2_handler; sigaction (SIGUSR2, &act, NULL);
  act.sa_handler = SIG_IGN;         sigaction (SIGPIPE, &act, NULL);
  act.sa_flags = SA_RESETHAND;
  act.sa_handler = sigterm_handler; sigaction (SIGINT , &act, NULL);
  act.sa_handler = sigterm_handler; sigaction (SIGTERM, &act, NULL);
}

int
main (int argc, char **argv, char **envp)
{
  ERR_load_crypto_strings (); // we have the RAM

  set_loglevel (L_INFO);
  set_identity (argv[0]);
  log_to (LOGTO_SYSLOG | LOGTO_STDERR);

  setlocale (LC_ALL, "");
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  parse_options (argc, argv, envp);

  argc -= optind;
  argv += optind;

  if (show_version)
    {
      printf (_("%s version %s (built %s %s, protocol version %d.%d)\n"), get_identity (),
              VERSION, __DATE__, __TIME__, PROTOCOL_MAJOR, PROTOCOL_MINOR);
      printf (_("Built with kernel interface %s/%s.\n"), IFTYPE, IFSUBTYPE);
      printf (_
              ("Copyright (C) 2003 Marc Lehmann <gvpe@schmorp.de> and others.\n"
               "See the AUTHORS file for a complete list.\n\n"
               "tinc comes with ABSOLUTELY NO WARRANTY.  This is free software,\n"
               "and you are welcome to redistribute it under certain conditions;\n"
               "see the file COPYING for details.\n"));

      return 0;
    }

  if (show_help)
    usage (0);

  log_to (LOGTO_SYSLOG | LOGTO_STDERR);

  /* Lock all pages into memory if requested */

#if HAVE_MLOCKALL && HAVE_SYS_MMAN_H && _POSIX_MEMLOCK
  if (do_mlock)
    if (mlockall (MCL_CURRENT | MCL_FUTURE))
      slog (L_ERR, _("system call `%s' failed: %s"), "mlockall", strerror (errno));
#endif

  if (argc >= 1)
    {
      thisnode = *argv++;
      argc--;
    }

  if (!ev_default_loop (0))
    {
      slog (L_ERR, _("unable to initialise the event loop (bad $LIBEV_METHODS?)"));
      exit (EXIT_FAILURE);
    }

  {
    configuration_parser (conf, true, argc, argv);
  }

  set_loglevel (llevel != L_NONE ? llevel : conf.llevel);

  RAND_load_file ("/dev/urandom", 1024);

  if (!THISNODE)
    {
      slog (L_ERR, _("current node not set, or node '%s' not found in configfile, specify the nodename when starting gvpe."),
            thisnode ? thisnode : "<unset>");
      exit (EXIT_FAILURE);
    }

  if (detach (do_detach))
    exit (EXIT_SUCCESS);

  setup_signals ();

  if (!network.setup ())
    {
      ev_loop (EV_DEFAULT_ 0);
      cleanup_and_exit (EXIT_FAILURE);
    }

  slog (L_ERR, _("unable to setup network, unrecoverable error, exiting."));
  cleanup_and_exit (EXIT_FAILURE);
}

